### PR TITLE
instancetype: Improve `PreferredInterfaceMasquerade` application 

### LINKED ISSUE #10053: `PreferredInterfaceMasquerade` applied to interfaces using `SRIOV` etc
/area instancetype
/sig network

**What happened**:

Our `PreferredInterfaceMasquerade` application code is pretty crude at the moment:

https://github.com/kubevirt/kubevirt/blob/534bda181a078a90215763f849eb4eabcb8d8da2/pkg/instancetype/instancetype.go#L1223-L1225

We really need input from networking folks to ensure this is only applied when valid.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
Steps to reproduce the behavior.

**Additional context**:

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by lyarwood**
/area instancetype
/cc @AlonaKaplan 
/cc @oshoval
/cherry-pick release-1.0

**What this PR does / why we need it**:

As set out in https://github.com/kubevirt/kubevirt/issues/10053 we shouldn't apply `PreferredInterfaceMasquerade`
when another binding is configured for an interface such as SRIOV etc.
We also shouldn't apply PreferredInterfaceMasquerade on non PodNetwork
interfaces. This change introduces basic checks for each of these and
adds simple unit test coverage.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10053

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/instancetype/instancetype.go [hunk 1]
```diff
@@ -1221,13 +1221,17 @@ func applyDiskPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePref
 	}
 }
 
+func isInterfaceBindingUnset(vmiIface *virtv1.Interface) bool {
+	return vmiIface.Masquerade == nil && vmiIface.Bridge == nil && vmiIface.Slirp == nil && vmiIface.SRIOV == nil && vmiIface.Macvtap == nil && vmiIface.Passt == nil
+}
+
```
Comment (oshoval) [inline pkg/instancetype/instancetype.go:1227]: Thanks, nice

Can we use something generic ? 
maybe something like this (didn't try it, just to show the idea)

```
func areAllFieldsNil(ifaceBindingMethod *InterfaceBindingMethod) bool {
	val := reflect.ValueOf(ifaceBindingMethod).Elem()
	return val.IsZero()
}
```

Comment (lyarwood) [reply pkg/instancetype/instancetype.go]: Yup sorry missed `InterfaceBindingMethod` was being used, updated.

Comment (oshoval) [reply pkg/instancetype/instancetype.go]: would panic if nil `panic: reflect: call of reflect.Value.IsZero on zero Value`
do we know it is not nil ?
otherwise we need please
```
	if ifaceBindingMethod == nil {
		return true
	}
```

Comment (lyarwood) [reply pkg/instancetype/instancetype.go]: I don't think it's ever `nil` when `inlined`. The existing test interfaces should cover this as neither have a binding mode set at the moment and we don't panic:

https://github.com/kubevirt/kubevirt/blob/178a70fc9203f6fa19e0bd6b7970f64429c42238/pkg/instancetype/instancetype_test.go#L1445-L1450

I'll start refactoring these tests shortly under the following issue FWIW, they really need some love:

https://github.com/kubevirt/kubevirt/issues/10175

Comment (oshoval) [reply pkg/instancetype/instancetype.go]: Ok thanks, and what about the `&` can it be removed ? or it is a must

Comment (lyarwood) [reply pkg/instancetype/instancetype.go]: It's required as you can't call `Elem` on a struct value, only Interface or Pointer :

`reflect: call of reflect.Value.Elem on struct Value`

File: pkg/instancetype/instancetype.go [hunk 2]
```diff
@@ -1221,13 +1222,17 @@ func applyDiskPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePref
...
 func applyInterfacePreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
 	for ifaceIndex := range vmiSpec.Domain.Devices.Interfaces {
 		vmiIface := &vmiSpec.Domain.Devices.Interfaces[ifaceIndex]
 		if preferenceSpec.Devices.PreferredInterfaceModel != "" && vmiIface.Model == "" {
 			vmiIface.Model = preferenceSpec.Devices.PreferredInterfaceModel
 		}
-		if preferenceSpec.Devices.PreferredInterfaceMasquerade != nil && vmiIface.Masquerade == nil {
+		if preferenceSpec.Devices.PreferredInterfaceMasquerade != nil && isInterfaceBindingUnset(vmiIface) {
```
Comment (oshoval) [inline pkg/instancetype/instancetype.go:1235]: Better to pass please only the param that we need
`vmiIface.InterfaceBindingMethod` (`&` is not needed IIUC)

**Review by oshoval — APPROVED** [review]
Thanks !

File: pkg/instancetype/instancetype.go [hunk 3]
```diff
@@ -1221,13 +1222,17 @@ func applyDiskPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePref
 	}
 }
 
+func isInterfaceBindingUnset(interfaceBindingMethod virtv1.InterfaceBindingMethod) bool {
+	return reflect.ValueOf(&interfaceBindingMethod).Elem().IsZero()
+}
+
 func applyInterfacePreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
 	for ifaceIndex := range vmiSpec.Domain.Devices.Interfaces {
```
[the remark this answers is not in the document]
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:1239]: Updated, appreciate this is documented in the user-guide but it might be useful if someone could also list these requirements out in the API docs?

https://kubevirt.io/api-reference/main/definitions.html#_v1_interface

**Review by oshoval — APPROVED** [review]

File: pkg/instancetype/instancetype_test.go [hunk 4]
```diff
@@ -1484,6 +1487,33 @@ var _ = Describe("Instancetype and Preferences", func() {
...
+					Expect(vmi.Spec.Domain.Devices.Interfaces[0].Masquerade).ToNot(BeNil())
+					Expect(vmi.Spec.Domain.Devices.Interfaces[1].Masquerade).To(BeNil())
+				})
+				It("should not be applied if primary interface has another binding set", func() {
+					vmi.Spec.Domain.Devices.Interfaces[0].SRIOV = &v1.InterfaceSRIOV{}
+					Expect(instancetypeMethods.ApplyToVmi(field, instancetypeSpec, preferenceSpec, &vmi.Spec)).To(BeNil())
+					Expect(vmi.Spec.Domain.Devices.Interfaces[0].Masquerade).To(BeNil())
+					Expect(vmi.Spec.Domain.Devices.Interfaces[0].SRIOV).ToNot(BeNil())
+				})
+				It("should not be applied if primary interface is not on Pod network", func() {
```
Comment (AlonaKaplan) [inline pkg/instancetype/instancetype_test.go:1508]: I think the terms primary and pod interface refer to the same, I would rephrase to "should not be applied on interface that is not on Pod network"

**Comment by AlonaKaplan** [comment]
/approve
Thanks!

File: pkg/instancetype/instancetype.go [hunk 5]
```diff
@@ -1221,13 +1222,26 @@ func applyDiskPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePref
 	}
 }
 
+func isInterfaceBindingUnset(interfaceBindingMethod virtv1.InterfaceBindingMethod) bool {
+	return reflect.ValueOf(&interfaceBindingMethod).Elem().IsZero()
```
Comment (0xFelix) [inline pkg/instancetype/instancetype.go:1226]: I think you can omit the additional ref and deref:

```
return reflect.ValueOf(interfaceBindingMethod).IsZero()
```

This should work too.

**Review by 0xFelix — APPROVED** [review]
/lgtm
